File: app-miyou/src/main/java/com/miyou/app/domain/credit/model/UserCredit.java
package com.miyou.app.domain.credit.model;

import com.miyou.app.domain.credit.exception.InsufficientCreditException;
import com.miyou.app.domain.dialogue.model.UserId;

public record UserCredit(
	UserId userId,
	long balance,
	long version
) {
	public UserCredit {
		if (userId == null) {
			throw new IllegalArgumentException("userId cannot be null");
		}
		if (balance < 0) {
			throw new IllegalArgumentException("balance cannot be negative");
		}
	}

	public UserCredit deduct(long amount) {
		if (balance < amount) {
			throw new InsufficientCreditException(userId, balance, amount);
		}
		return new UserCredit(userId, balance - amount, version);
	}

	public UserCredit charge(long amount) {
		if (amount <= 0) {
			throw new IllegalArgumentException("charge amount must be positive");
		}
		return new UserCredit(userId, balance + amount, version);
	}

	public static UserCredit initialize(UserId userId, long initialBalance) {
		return new UserCredit(userId, initialBalance, 0L);
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/credit/model/UserCredit.kt
package com.miyou.app.domain.credit.model

import com.miyou.app.domain.credit.exception.InsufficientCreditException

data class UserCredit(
    val userId: String,
    val balance: Long,
    val version: Long,
) {
    init {
        require(balance >= 0) { "balance cannot be negative" }
    }

    fun deduct(amount: Long): UserCredit {
        require(amount > 0) { "deduct amount must be positive" }
        if (balance < amount) {
            throw InsufficientCreditException(userId, balance, amount)
        }
        return UserCredit(userId, balance - amount, version)
    }

    fun charge(amount: Long): UserCredit {
        require(amount > 0) { "charge amount must be positive" }
        return UserCredit(userId, balance + amount, version)
    }

    companion object {
        @JvmStatic
        fun initialize(
            userId: String,
            initialBalance: Long,
        ): UserCredit {
            require(initialBalance >= 0) { "initial balance cannot be negative" }
            return UserCredit(userId, initialBalance, 0L)
        }
    }

    fun userId(): String = userId

    fun balance(): Long = balance

    fun version(): Long = version
}
